src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt

```diff
@@ -281,10 +270,18 @@ class ReferralService(
       )
     }
 
-    update.serviceCategoryIds?.let {
-      referral.selectedServiceCategories = serviceCategoryRepository.findByIdIn(it)
+    update.serviceCategoryIds?.let { serviceCategoryIds ->
+      referral.complexityLevelIds = referral.complexityLevelIds?.filter { complexityLevel -> serviceCategoryIds.contains(complexityLevel.key) }?.toMutableMap()
+      referral.selectedDesiredOutcomes = referral.selectedDesiredOutcomes?.filter { desiredOutcome -> serviceCategoryIds.contains(desiredOutcome.serviceCategoryId) }?.toMutableList()
     }
 
+    // Need to save and flush the entity before removing selected service categories due to constraint violations being thrown
+    referralRepository.save(referral)
+    referralRepository.flush()
+
+    update.serviceCategoryIds?.let { serviceCategoryIds ->
+      referral.selectedServiceCategories = serviceCategoryRepository.findByIdIn(serviceCategoryIds)
+    }
     return referralRepository.save(referral)
   }
```

Comment:
Need to save, flush and save to prevent the entitymanager attempting to delete the selected service categories first before the selected complexity level and desired outcomes. This would otherwise throw a constraint violation.

(This occurs even if you order sequence of statements in the correct order)

Reply:
Would this be simpler if both our data model and the JSON payload would support this on a separate endpoints?

`PUT /draft-referral/{id}/service-category` &rarr; creates or updates categories, outcomes, complexities
`DELETE /draft-referral/{id}/service-category` &rarr; deletes

Reply:
Some good ideas..
However I think the way that the DraftReferral has been designed is that complexity level and desired outcomes are not directly linked to the service category and are instead linked to the referral so it would be a bit less confusing if the update occurred on the DraftReferral resources directly. The user will never want to DELETE a selected ServiceCategory instead they would only ever want to replace it with a new selection.